### PR TITLE
Process info standard input encoding

### LINKED ISSUE #20497: Add ProcessStartInfo.StandardInputEncoding property
See **detailed latest proposal** in https://github.com/dotnet/corefx/issues/20497#issuecomment-312896302

# Original Proposal

By default, the standard IO readers and writer for a process use `Console.OutputEncoding` and `Console.InputEncoding` respectively. 

The `ProcessStartInfo` class provides `StandardOutputEncoding` and `StandardErrorEncoding` properties to allow callers to override the encoding for those. This simply changes the encoding passed to the `StreamReader` object which wraps the underlying process byte stream.

To more naturally support talking to a process that uses a different encoding than the current process's console encoding, we should add a new property to `ProcessStartInfo`:

```
public Encoding StandardInputEncoding { get; set; }
```

This would have the same semantics as the existing properties:
* Defaults to `null`, which means use the default encoding
* If specified and the input stream is not redirected, an exception is thrown on process start

[EDIT] Added latest proposal link by @karelz

### PR DESCRIPTION
**Pull request by madelson**
Implement `ProcessStartInfo.StandardInputEncoding` property. Fix #20497

### CODE REVIEW

File: src/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests.csproj [hunk 1]
```diff
@@ -34,10 +34,12 @@
     <Compile Include="ProcessStandardConsoleTests.cs" />
     <Compile Include="ProcessStartInfoTests.cs" />
     <Compile Include="ProcessStreamReadTests.cs" />
+    <Compile Condition="'$(TargetGroup)' == 'netcoreapp' or '$(TargetGroup)' == 'uap'" Include="ProcessStreamReadTests.netcoreapp2.1.cs" />
```
Comment (safern) [inline src/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests.csproj]: Could you please move this two Conditioned Compile items to a separate ItemGroup and condition the ItemGroup instead? That makes it much cleaner to read.

File: src/System.Diagnostics.Process/tests/ProcessStreamReadTests.netcoreapp2.1.cs [hunk 2]
```diff
@@ -0,0 +1,33 @@
+﻿using System;
```
Comment (safern) [inline src/System.Diagnostics.Process/tests/ProcessStreamReadTests.netcoreapp2.1.cs]: I think I mentioned to call it netcoreapp2.1 but we are trying to go with the versionless convention. Could you please rename this two files to .netcoreapp.cs versionless?

Comment (danmosemsft) [inline src/System.Diagnostics.Process/tests/ProcessStreamReadTests.netcoreapp2.1.cs]: Please add the usual 3 line license header

File: src/System.Diagnostics.Process/tests/ProcessStreamReadTests.netcoreapp2.1.cs [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+        public void TestCustomStandardInputEncoding()
+        {
+            var process = CreateProcessPortable(RemotelyInvokable.ReadLineWithCustomEncodingWriteLineWithUtf8, Encoding.UTF32.WebName);
+            process.StartInfo.RedirectStandardInput = true;
+            process.StartInfo.StandardInputEncoding = Encoding.UTF32;
+            process.StartInfo.RedirectStandardOutput = true;
+            process.StartInfo.StandardOutputEncoding = Encoding.UTF8;
+            process.Start();
+
+            const string TestLine = "\U0001f627\U0001f62e\U0001f62f";
```
Comment (danmosemsft) [inline src/System.Diagnostics.Process/tests/ProcessStreamReadTests.netcoreapp2.1.cs]: Can you add a negative test, where SIE is set to some other encoding, and the output does not match the expected value?

File: src/System.Diagnostics.Process/tests/RemotelyInvokable.cs [hunk 4]
```diff
@@ -111,6 +112,22 @@ public static int ReadLineWriteIfNull()
             return SuccessExitCode;
         }
 
+        public static int ReadLineWithCustomEncodingWriteLineWithUtf8(string inputEncoding)
```
Comment (danmosemsft) [inline src/System.Diagnostics.Process/tests/RemotelyInvokable.cs]: With a little refactoring you could make this more generic `ReadAndWriteLineWithEncoding(string, string)` ?

Comment (danmosemsft) [reply src/System.Diagnostics.Process/tests/RemotelyInvokable.cs]: maybe not worth the bother

Comment (madelson) [reply src/System.Diagnostics.Process/tests/RemotelyInvokable.cs]: @danmosemsft sure, but is that useful/worthwhile here? This function exists solely to support the two test methods that don't need that additional genericity.

Comment (danmosemsft) [reply src/System.Diagnostics.Process/tests/RemotelyInvokable.cs]: nope...

**Comment by danmosemsft** [comment]
@madelson as you see from the UWP failure logs, you need to add your function to RemotelyInvokable
```
System.Exception : Method ReadLineWithCustomEncodingWriteLineWithUtf8UapCmd could not be found in class RemotelyInvokable.
Stack Trace :
   at System.Diagnostics.Tests.ProcessTestBase.GetMethodForUap(MethodInfo originalMethod) in D:\j\workspace\windows-TGrou---33cbf18b\src\System.Diagnostics.Process\tests\ProcessTestBase.Uap.cs:line 53
   at System.Diagnostics.Tests.ProcessTestBase.CreateProcessForUap(MethodInfo method, String[] args) in D:\j\workspace\windows-TGrou---33cbf18b\src\System.Diagnostics.Process\tests\ProcessTestBase.Uap.cs:line 68
   at System.Diagnostics.Tests.ProcessTestBase.CreateProcessForUap(Func`2 func, String arg) in D:\j\workspace\windows-TGrou---33cbf18b\src\System.Diagnostics.Process\tests\ProcessTestBase.Uap.cs:line 39
   at System.Diagnostics.Tests.ProcessTestBase.CreateProcessPortable(Func`2 func, String arg) in D:\j\workspace\windows-TGrou---33cbf18b\src\System.Diagnostics.Process\tests\ProcessTestBase.Uap.cs:line 29
   at System.Diagnostics.Tests.ProcessStreamReadTests.TestCustomStandardInputEncoding() in D:\j\workspace\windows-TGrou---33cbf18b\src\System.Diagnostics.Process\tests\ProcessStreamReadTests.netcoreapp.cs:line 19
```

Although @safern I am confused as to the purpose of the RemotelyInvokable class. It suggests it is for UAP, but some of its methods should actually work these days in UAP.  It also mentions NETFX. Plus many non UAP tests use it.

**Comment by safern** [comment]
I'm also confused now. @krwq who added RemotelyInvokable class.

**Comment by krwq** [comment]
RemotelyInvokable was created because at the time RemoteInvoke on UAP has not returned the Process which is used by most of the tests and thus the tests were disabled. The only way to change that would be to update RemoteInvoke so that the other process calls winapi's GetProcessId upon start, communicate it back to the host process the open the process handle and then create Process class by private reflection injecting with the handle you obtained - even with that in place there were some tests which did not work correctly as they were expecting the process to start supsended so the communication upon start was not always a good option. I've noted that RemoteInvoke was used only to simple things like printing a line and similar so I've implemented corresponding version of that which only uses cmd.exe to write lines and thus we could create a process which does what we want. Because of that limitation now adding code which needs to be remotely invoked requires adding two versions of the code - one for cmd (with UapCmd suffix) and one "regular".

If currently there is no such limitation this code can be safely removed otherwise we will be without test coverage on UAP.

**Comment by madelson** [comment]
@danmosemsft @safern @krwq I'm not sure what the next step is here. Looking at RemotelyInvokable, it seems like for UAP the idea is to replicate the C# logic with a set of CMD commands. Not being a CMD expert, it's not obvious to me how to replicate my encodings test case in this form. Any ideas?

Also, I noticed that for one case we have:
```
        public static string WriteSlowlyByByteUapCmd()
        {
            throw new Exception("No simple way of doing this using cmd.exe");
        }
```

Would this be an acceptable solution?

**Comment by krwq** [comment]
@madelson yes but please ensure that the test is also disabled on UAP

**Comment by madelson** [comment]
@krwq the preferred way to do this would be to have `ProcessStreamReadTests.netcoreapp.cs` lose the `or '$(TargetGroup)' == 'uap'` part of it's `Condition` in the csproj file? Do I have to do both that *and* add the `ReadLineWithCustomEncodingWriteLineWithUtf8UapCmd` stub method?

**Comment by krwq** [comment]
@madelson we usually do it with Xunit attribute like here:
https://github.com/dotnet/corefx/blob/c51694a1065e40f6c92d1fb8d82f3d879cb3d4ad/src/System.Diagnostics.Process/tests/ProcessTests.cs#L61

If runs without UapCmd method then remove it - I think I might have added the check on both version but do not remember right now.

**Comment by krwq** [comment]
Thank you @madelson!

**Comment by madelson** [comment]
Anything further required from my end here?

**Comment by danmosemsft** [comment]
Thanks!

**Comment by khellang** [comment]
good job

**Pull request merged**